File: Problem/By section/Greedy/Problem 53/problem_53.cpp
#include <bits/stdc++.h>

using namespace std;

// {-2, 1, -3, 4, -1, 2, 1, -5, 4}
// {1, -3}

class Solution {
public:
    // Time: O(n), Space: O(1)
    int maxSubArray(vector<int>& nums) {
        int max_sum = INT_MIN;
        int temp = INT_MIN;

        for(auto& i : nums) {
            if(temp < 0) {
                temp = i;
            }else{
                temp += i;
            }

            if(max_sum < temp) {
                max_sum = temp;
            }
        }

        return max_sum;
    }
};

int main(int argc, char** argv) {
    Solution s;
    vector<int> input = {5, 4, -1, 7, 8};
    cout << s.maxSubArray(input) << endl;
    return 0;
}

File: Problem/By section/Greedy/Problem 53/problem_53.java
class Solution {
    // Time: O(n), Space: O(1)
    public int maxSubArray(int[] nums) {
        if(nums.length == 0) return 0;
        
        int curr_sum = nums[0];
        int ans = curr_sum;
        
        // iterate through the whole array
        // we know that if the 'curr_sum' < 0, it cannot be part of the max sum
        // and we update the 'ans' every time to ensure it can record all the max sum
        for(int i = 1; i < nums.length; i++){
            curr_sum = (curr_sum > 0) ? curr_sum + nums[i] : nums[i];
            ans = Math.max(ans, curr_sum);
        }
        
        return ans;
    }
}